token-upgrade/program/src/instruction.rs

```diff
@@ -0,0 +1,30 @@
+//! Program instructions
+
+use num_enum::{IntoPrimitive, TryFromPrimitive};
+
+/// Instructions supported by the TokenUpgrade program
+#[derive(Clone, Copy, Debug, TryFromPrimitive, IntoPrimitive)]
+#[repr(u8)]
+pub enum TokenUpgradeInstruction {
+    /// Burns all of the source tokens in the user's account, and transfers the same
+    /// amount of tokens from an account owned by a PDA into another account.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[writeable]` Source token account to burn from
+    ///   1. `[writeable]` Source token mint
+    ///   2. `[writeable]` Destination token upgrade account, owned by:
+    ///       `get_token_upgrade_authority_address(source_mint, destination_mint, program_id)`
+    ///   3. `[writeable]` Destination token account to transfer into
+    ///   4. `[]` Destination token mint
+    ///   5. `[signer]` Source token account transfer authority (owner or delegate)
+    ///   6. `[]` PDA owner of token upgrade account, given by:
+    ///       `get_token_upgrade_authority_address(source_mint, destination_mint, program_id)`
```

Comment:
```suggestion
    ///   6. `[]` Transfer authority (owner or delegate) of destination tokens held in program escrow, must be:
    ///       `get_token_upgrade_authority_address(source_mint, destination_mint, program_id)`
```

Specifically I think this'll still work if the program is only a delegate, which gives the owner a nice way to yank back any remaining tokens from the program if ever needed

Reply:
great idea!